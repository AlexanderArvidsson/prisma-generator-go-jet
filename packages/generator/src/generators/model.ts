import { DMMF } from '@prisma/generator-helper'
import path from 'path'
import { OUTPUT_HEADER } from '../constants.js'
import { GeneratorConfig } from '../generator.js'
import { formatField, getModelName } from '../helpers/format.js'
import { SchemaBuilder } from '../helpers/schema.js'
import { writeFileSafely } from '../utils/writeFileSafely.js'

export type ModelField = {
  name: string
  type: string
  tags: string[]
}

export type RenderModelTemplateOptions = {
  packageName: string
  modelName: string
  imports: string[]
  fields: ModelField[]
}

export function renderModelTemplate({
  packageName,
  modelName,
  imports,
  fields,
}: RenderModelTemplateOptions) {
  let importsStr = ''

  if (imports.length > 0) {
    importsStr = `import (
  ${imports.map((i) => `"${i}"`).join('\n  ')}
)`
  }

  return `package ${packageName}

${importsStr}

type ${modelName} struct {
  ${fields
    .map(
      ({ name, type, tags }) =>
        `${name} ${type}` + (tags.length > 0 ? ` \`${tags.join(' ')}\`` : ''),
    )
    .join('\n  ')}
}`
}

export function renderModel(
  model: DMMF.Model,
  builder: SchemaBuilder,
  packageName: string,
) {
  const modelName = getModelName(model)

  const schemaModel = builder.findByType('model', { name: model.name })
  if (!schemaModel) throw new Error(`Model ${modelName} not found in schema`)

  const imports: Set<string> = new Set()
  const fields = model.fields
    .filter((field) => !field.relationName)
    .map((field) => {
      const { name, typeImport, type, tags } = formatField(
        field,
        schemaModel,
        builder,
      )

      if (typeImport) imports.add(typeImport)

      return {
        name,
        type,
        tags,
      }
    })

  return renderModelTemplate({
    packageName,
    modelName,
    imports: Array.from(imports),
    fields,
  })
}

export async function generateModel(
  model: DMMF.Model,
  config: GeneratorConfig,
  builder: SchemaBuilder,
  output: string,
) {
  const modelName = getModelName(model)

  const location = path.join(
    output,
    config.outputModels,
    modelName.toLowerCase() + '.go',
  )
  const packageName = path.basename(path.dirname(location))

  const content = renderModel(model, builder, packageName)

  await writeFileSafely(location, content, {
    header: OUTPUT_HEADER,
  })
}
