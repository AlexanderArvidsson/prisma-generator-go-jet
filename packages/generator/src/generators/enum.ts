import { DMMF } from '@prisma/generator-helper'
import { getEnumMemberName, getEnumName } from '../helpers/format.js'
import { GeneratorConfig } from '../generator.js'
import { OUTPUT_HEADER } from '../constants.js'
import { writeFileSafely } from '../utils/writeFileSafely.js'
import path from 'path'

export type EnumMember = {
  name: string
  value: string
}

export type RenderEnumTemplateOptions = {
  packageName: string
  enumName: string
  members: EnumMember[]
}

export function renderEnumTemplate({
  packageName,
  enumName,
  members,
}: RenderEnumTemplateOptions) {
  return `package ${packageName}

import "github.com/go-jet/jet/v2/postgres"

var ${enumName} = &struct {
  ${members.map((member) => `${member.name} postgres.StringExpression`).join('\n  ')}
}{
  ${members.map((member) => `${member.name}: postgres.NewEnumValue("${member.value}"),`).join('\n  ')}
}`
}

export function renderEnum(
  datamodelEnum: DMMF.DatamodelEnum,
  packageName: string,
) {
  const enumName = getEnumName(datamodelEnum)

  const members = datamodelEnum.values.map(
    (member): EnumMember => ({
      name: getEnumMemberName(member.name),
      value: member.name,
    }),
  )

  return renderEnumTemplate({
    packageName,
    enumName,
    members,
  })
}

export async function generateEnum(
  datamodelEnum: DMMF.DatamodelEnum,
  config: GeneratorConfig,
  output: string,
) {
  const enumName = getEnumName(datamodelEnum)

  const location = path.join(
    output,
    config.outputEnums,
    enumName.toLowerCase() + '.go',
  )
  const packageName = path.basename(path.dirname(location))

  const content = renderEnum(datamodelEnum, packageName)

  await writeFileSafely(location, content, { header: OUTPUT_HEADER })
}
