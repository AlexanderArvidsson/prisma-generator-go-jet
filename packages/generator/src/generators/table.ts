import { DMMF, ReadonlyDeep } from '@prisma/generator-helper'
import path from 'path'
import { GeneratorConfig } from '../generator.js'
import { OUTPUT_HEADER } from './../constants.js'
import { formatField, getModelName } from './../helpers/format.js'
import { SchemaBuilder } from './../helpers/schema.js'
import { pgTypeMap } from './../helpers/types.js'
import { writeFileSafely } from './../utils/writeFileSafely.js'

export type TableColumn = {
  name: string
  columnName: string
  type: string
  mutable?: boolean
}

export type RenderTableTemplateOptions = {
  packageName: string
  modelName: string
  schemaName: string
  alias?: string
  tableName: string
  privateModelName: string
  columns: TableColumn[]
}

export function renderTableTemplate({
  packageName,
  modelName,
  schemaName,
  tableName,
  alias = '',
  privateModelName,
  columns,
}: RenderTableTemplateOptions) {
  return `package ${packageName}

import (
	"github.com/go-jet/jet/v2/postgres"
)

var ${modelName} = new${modelName}Table("${schemaName}", "${tableName}", "${alias}")

type ${privateModelName}Table struct {
	postgres.Table

	// Columns
  ${columns.map((c) => `${c.name} postgres.Column${c.type}`).join('\n  ')}

	AllColumns     postgres.ColumnList
	MutableColumns postgres.ColumnList
}

type ${modelName}Table struct {
	${privateModelName}Table

	EXCLUDED ${privateModelName}Table
}

// AS creates new ${modelName}Table with assigned alias
func (a ${modelName}Table) AS(alias string) *${modelName}Table {
	return new${modelName}Table(a.SchemaName(), a.TableName(), alias)
}

// Schema creates new ${modelName}Table with assigned schema name
func (a ${modelName}Table) FromSchema(schemaName string) *${modelName}Table {
	return new${modelName}Table(schemaName, a.TableName(), a.Alias())
}

// WithPrefix creates new ${modelName}Table with assigned table prefix
func (a ${modelName}Table) WithPrefix(prefix string) *${modelName}Table {
	return new${modelName}Table(a.SchemaName(), prefix+a.TableName(), a.TableName())
}

// WithSuffix creates new ${modelName}Table with assigned table suffix
func (a ${modelName}Table) WithSuffix(suffix string) *${modelName}Table {
	return new${modelName}Table(a.SchemaName(), a.TableName()+suffix, a.TableName())
}

func new${modelName}Table(schemaName, tableName, alias string) *${modelName}Table {
	return &${modelName}Table{
		${privateModelName}Table: new${modelName}TableImpl(schemaName, tableName, alias),
		EXCLUDED: new${modelName}TableImpl("", "excluded", ""),
	}
}

func new${modelName}TableImpl(schemaName, tableName, alias string) ${privateModelName}Table {
	var (
    ${columns.map((column) => `${column.name}Column = postgres.${column.type}Column("${column.columnName}")`).join('\n    ')}
		allColumns                     = postgres.ColumnList{${columns.map((column) => `${column.name}Column`).join(', ')}}
		mutableColumns                 = postgres.ColumnList{${columns
      .filter((column) => column.mutable)
      .map((column) => `${column.name}Column`)
      .join(', ')}}
	)

	return ${privateModelName}Table{
		Table: postgres.NewTable(schemaName, tableName, alias, allColumns...),

		//Columns
    ${columns.map((column) => `${column.name}: ${column.name}Column,`).join('\n    ')}

		AllColumns:     allColumns,
		MutableColumns: mutableColumns,
	}
}`
}

export function renderTable(
  model: DMMF.Model,
  builder: SchemaBuilder,
  schemaName: string,
  packageName: string,
) {
  const modelName = getModelName(model)
  const privateModelName =
    modelName.charAt(0).toLowerCase() + modelName.slice(1)

  const tableName = model.dbName?.toString() ?? model.name

  const schemaModel = builder.findByType('model', { name: model.name })
  if (!schemaModel) throw new Error(`Model ${modelName} not found in schema`)

  const columns = model.fields
    .filter((field) => !field.relationName)
    .map((field): TableColumn => {
      const { name } = formatField(field, schemaModel, builder)

      return {
        name,
        columnName: field.dbName ?? field.name,
        type: pgTypeMap[field.type] ?? pgTypeMap.Unsupported,
        mutable: !field.isId && !field.isGenerated,
      }
    })

  return renderTableTemplate({
    modelName,
    privateModelName,
    schemaName,
    tableName,
    alias: modelName,
    packageName,
    columns,
  })
}

export async function generateTable(
  model: DMMF.Model,
  config: GeneratorConfig,
  builder: SchemaBuilder,
  output: string,
) {
  const modelName = getModelName(model)

  const location = path.join(
    output,
    config.outputTables,
    modelName.toLowerCase() + '.go',
  )
  const packageName = path.basename(path.dirname(location))

  const content = renderTable(model, builder, config.schemaName, packageName)

  await writeFileSafely(location, content, { header: OUTPUT_HEADER })
}

export function renderTableUseSchema(
  models: ReadonlyDeep<DMMF.Model[]>,
  packageName: string,
) {
  const modelsStr = models
    .map((model) => {
      const modelName = getModelName(model)
      return `  ${modelName} = ${modelName}.FromSchema(schema)`
    })
    .join('\n')

  return `package ${packageName}

// UseSchema sets a new schema name for all generated table SQL builder types. It is recommended to invoke
// this method only once at the beginning of the program.
func UseSchema(schema string) {
${modelsStr}
}`
}

export async function generateTableUseSchema(
  models: ReadonlyDeep<DMMF.Model[]>,
  config: GeneratorConfig,
  output: string,
) {
  const location = path.join(output, config.outputTables, 'table_use_schema.go')
  const packageName = path.basename(path.dirname(location))

  const content = renderTableUseSchema(models, packageName)

  await writeFileSafely(location, content, { header: OUTPUT_HEADER })
}
