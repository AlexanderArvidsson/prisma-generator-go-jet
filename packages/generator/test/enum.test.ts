import { describe, expect, it } from 'vitest'
import { getSampleDMMF } from './__fixtures__/getSampleDMMF'
import { renderEnum } from '../src/generators/enum'

describe('enum generation', () => {
  it('generates all dmmf enums', async () => {
    const sampleDMMF = await getSampleDMMF()

    sampleDMMF.datamodel.enums.forEach((enumInfo) => {
      expect(renderEnum(enumInfo, 'enum')).toMatchSnapshot(enumInfo.name)
    })
  })
})
