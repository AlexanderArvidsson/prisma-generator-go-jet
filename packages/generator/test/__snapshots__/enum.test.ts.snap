// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`enum generation > generates all dmmf enums > Language 1`] = `
"package enum

import "github.com/go-jet/jet/v2/postgres"

var Language = &struct {
  Typescript postgres.StringExpression
  Javascript postgres.StringExpression
  Rust postgres.StringExpression
  Go postgres.StringExpression
  Python postgres.StringExpression
  Cpp postgres.StringExpression
}{
  Typescript: postgres.NewEnumValue("Typescript"),
  Javascript: postgres.NewEnumValue("Javascript"),
  Rust: postgres.NewEnumValue("Rust"),
  Go: postgres.NewEnumValue("Go"),
  Python: postgres.NewEnumValue("Python"),
  Cpp: postgres.NewEnumValue("Cpp"),
}"
`;

exports[`enum generation > generates all dmmf enums > NotificationType 1`] = `
"package enum

import "github.com/go-jet/jet/v2/postgres"

var NotificationType = &struct {
  NewPosts postgres.StringExpression
  NewComments postgres.StringExpression
  NewFollowers postgres.StringExpression
  Reply postgres.StringExpression
  HeartOnPost postgres.StringExpression
  HeartOnComment postgres.StringExpression
  HeartOnReply postgres.StringExpression
}{
  NewPosts: postgres.NewEnumValue("newPosts"),
  NewComments: postgres.NewEnumValue("newComments"),
  NewFollowers: postgres.NewEnumValue("newFollowers"),
  Reply: postgres.NewEnumValue("reply"),
  HeartOnPost: postgres.NewEnumValue("heartOnPost"),
  HeartOnComment: postgres.NewEnumValue("heartOnComment"),
  HeartOnReply: postgres.NewEnumValue("heartOnReply"),
}"
`;

exports[`enum generation > generates all dmmf enums > State 1`] = `
"package enum

import "github.com/go-jet/jet/v2/postgres"

var State = &struct {
  Active postgres.StringExpression
  Pending postgres.StringExpression
  Banned postgres.StringExpression
}{
  Active: postgres.NewEnumValue("Active"),
  Pending: postgres.NewEnumValue("Pending"),
  Banned: postgres.NewEnumValue("Banned"),
}"
`;
